cub/test/catch2_test_device_reduce_by_key_iterators.cu

```diff
@@ -0,0 +1,125 @@
+ *
+ ******************************************************************************/
+
+#include <cub/device/device_reduce.cuh>
+
+#include <thrust/device_vector.h>
+#include <thrust/host_vector.h>
+#include <thrust/iterator/constant_iterator.h>
+
+#include <cstdint>
+
+#include "catch2_test_device_reduce.cuh"
+
+// Has to go after all cub headers. Otherwise, this test won't catch unused
+// variables in cub kernels.
+#include "c2h/custom_type.cuh"
+#include "catch2/catch.hpp"
+#include "catch2_test_cdp_helper.h"
+#include "catch2_test_helper.h"
+
+DECLARE_CDP_WRAPPER(cub::DeviceReduce::ReduceByKey, device_reduce_by_key);
+
+// %PARAM% TEST_CDP cdp 0:1
+
+// List of types to test
+using custom_t = c2h::custom_type_t<c2h::accumulateable_t, c2h::equal_comparable_t>;
+using iterator_type_list =
+  c2h::type_list<type_triple<custom_t>, type_triple<std::int64_t, std::int64_t, custom_t>>;
+
+CUB_TEST("Device reduce-by-key works with iterators", "[reduce][device]", iterator_type_list)
+{
+  using params   = params_t<TestType>;
+  using item_t   = typename params::item_t;
+  using output_t = typename params::output_t;
+  using key_t    = typename params::type_pair_t::key_t;
+  using offset_t = uint32_t;
+
+  constexpr offset_t min_items = 1;
+  constexpr offset_t max_items = 1000000;
+
+  // Number of items
+  const offset_t num_items = GENERATE_COPY(take(2, random(min_items, max_items)),
+                                           values({
+                                             min_items,
+                                             max_items,
+                                           }));
+  INFO("Test num_items: " << num_items);
+
+  // Range of segment sizes to generate (a segment is a series of consecutive equal keys)
+  const std::tuple<offset_t, offset_t> seg_size_range =
+    GENERATE_COPY(table<offset_t, offset_t>({{1, 1}, {1, num_items}, {num_items, num_items}}));
+  INFO("Test seg_size_range: [" << std::get<0>(seg_size_range) << ", "
+                                << std::get<1>(seg_size_range) << "]");
+
+  // Generate input segments
+  thrust::device_vector<offset_t> segment_offsets =
+    c2h::gen_uniform_offsets<offset_t>(CUB_SEED(1),
+                                       num_items,
+                                       std::get<0>(seg_size_range),
+                                       std::get<1>(seg_size_range));
+
+  // Get array of keys from segment offsets
+  const offset_t num_segments               = segment_offsets.size() - 1;
+  thrust::device_vector<key_t> segment_keys = c2h::get_key_segments<key_t>(segment_offsets,
+                                                                           num_items);
+  auto d_keys_it                            = thrust::raw_pointer_cast(segment_keys.data());
+
+  // Prepare input data
+  item_t default_constant{};
+  init_default_constant(default_constant);
+  auto in_it = thrust::make_constant_iterator(default_constant);
```

Comment:
Do you think a consistent naming `key_x` / `val_x` would be better?

Reply:
Agreed. Renamed the variables. Hope it reads a bit clearer now. Feel free to add further suggestions if you find something could still be improved further.